### PR TITLE
types of encoding and conclusion

### PR DESCRIPTION
**Pull request by leonardoalt**

### CODE REVIEW

**Comment by leonardoalt** [comment]
@chriseth can you review? This is the types of encoding thing you mentioned yesterday

File: conclusions.tex [hunk 1]
```diff
@@ -1,11 +1,16 @@
 \section{Conclusion}
 \label{section:conclusion}
+We have presented our work in progress building an SMT-based formal
+verification module inside the Solidity compiler.
+%
+The component creates SMT constraints from the Solidity code and queries to
+statically check for underflow/overflow, unreachable/trivial code, and
```
Comment (chriseth) [inline conclusions.tex]: Not to forget division by zero :)

File: conclusions.tex [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
-This is more complicated, because typically, the smart contract is not designed
-to assume all possible states over its potential lifetime. Therefore, the
-system should be able to restrict the set of possible states by analyzing
-its code.
-
-Apart from these explicit conditions, other conditions like overflows,
-divisions by zero or array access out of bounds (planned) are checked.
+Our under implementation features intend to extend the subset of Solidity that
+is supported, and future work includes interesting research questions, such as
+computing multi-transaction invariants for state variables in smart contracts,
```
Comment (chriseth) [inline conclusions.tex]: We should also mention modifiers and deriving postconditions from asserts mentioned at the end of function calls (or inside modifiers) which can then be used for generic invariants about storage variables.

File: smt_fv.tex [hunk 3]
```diff
@@ -14,46 +14,57 @@ \section{SMT-based Solidity verification}
...
+\end{paragraph}
+
+\begin{paragraph}{Variable assignment.}
 The encoding of variable assignment follows the \emph{Single Static Assignment}
 (SSA) where each assignment to a program variable introduces a new SMT variable
 that is assigned to only once.
 %
-The encoder currently supports \emph{Bool} and the various sizes of
-\emph{Integer} variables.
+When a program variable is modified inside a branch block (if-else), the
```
Comment (chriseth) [inline smt_fv.tex]: Perhaps "is modified at a place that is executed only conditionally (if-else), ..."?

File: smt_fv.tex [hunk 4]
```diff
@@ -76,27 +87,26 @@ \subsection{SMT encoding}
 %
 Failing calls to \code{require} are not treated as errors and
 are just checked for triviality and reachability.
-%
+\end{paragraph}
```
Comment (chriseth) [inline smt_fv.tex]: Perhaps we should add a paragraph here specifying how the various encodings above are combined:

To check an assertion, we use `$cf \land vars \land \neg ass$`, where `cf` is the control flow stack that reaches the assertion, `vars` are the SSA variables, expressions, their initial values and range contraints and `ass` is the condition in the assertion.

Hm, now I realize that we did not really talk about operators yet, did we?

File: conclusions.tex (hunk 2 above)
Comment (leonardoalt) [reply conclusions.tex]: Right. Maybe in the discussion section?

File: smt_fv.tex (hunk 3 above)
Comment (leonardoalt) [reply smt_fv.tex]: I'll try that

File: smt_fv.tex (hunk 4 above)
Comment (leonardoalt) [reply smt_fv.tex]: Yes, at some point I sort of try to make the distinction between the SMT encoding of the program, which is the type contraints, control flow, etc, versus an SMT query, which comes from targets. Maybe it makes sense to be more explicit about it.

We didn't explicitly talk about operators, but to me that's implicit in supporting types. I mean, if you support Integers, you support the integer operations, and not necessarily bit-wise operations, for example

Comment (leonardoalt) [reply smt_fv.tex]: Added some text in the beginning of the subsection

File: smt_fv.tex (hunk 3 above)
Comment (leonardoalt) [reply smt_fv.tex]: Swapped places, I think it worked

File: conclusions.tex (hunk 2 above)
Comment (leonardoalt) [reply conclusions.tex]: Added a sentence in the conclusions and a TODO for the discussion

**Comment by chriseth** [comment]
Let's merge this and I will take another look at the result.

**Pull request merged**